src/mds/QuiesceDbEncoding.h

```diff
@@ -0,0 +1,157 @@
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2023 IBM, Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+#include "QuiesceDb.h"
+#include "include/encoding.h"
+#include <stdint.h>
+
+void encode(QuiesceDbVersion const& v, bufferlist& bl, uint64_t features = 0)
+{
+  encode(v.epoch, bl, features);
+  encode(v.set_version, bl, features);
+}
+
+void decode(QuiesceDbVersion& v, bufferlist::const_iterator& p)
+{
+  decode(v.epoch, p);
+  decode(v.set_version, p);
+}
+
+void encode(QuiesceState const & state, bufferlist& bl, uint64_t features=0)
+{
+  static_assert(QuiesceState::QS__MAX <= UINT8_MAX);
+  uint8_t v = (uint8_t)state;
+  encode(v, bl, features);
+}
+
+void decode(QuiesceState & state, bufferlist::const_iterator& p)
+{
+  uint8_t v = 0;
+  decode(v, p);
+  state = (QuiesceState)v;
+}
+
+void encode(QuiesceTimeInterval const & interval, bufferlist& bl, uint64_t features=0)
+{
+  encode(interval.count(), bl, features);
+}
+
+void decode(QuiesceTimeInterval & interval, bufferlist::const_iterator& p)
+{
+  QuiesceClock::rep count;
+  decode(count, p);
+  interval = QuiesceTimeInterval { count };
+}
+
+void encode(RecordedQuiesceState const& rstate, bufferlist& bl, uint64_t features = 0)
+{
+  encode(rstate.state, bl, features);
+  encode(rstate.at_age.count(), bl, features);
+}
+
+void decode(RecordedQuiesceState& rstate, bufferlist::const_iterator& p)
+{
+  decode(rstate.state, p);
+  decode(rstate.at_age, p);
+}
+
+void encode(QuiesceSet::MemberInfo const& member, bufferlist& bl, uint64_t features = 0)
+{
+  encode(member.rstate, bl, features);
+  encode(member.excluded, bl, features);
```

Comment:
I didn't want to do that with the internal classes, otherwise, we'd waste lots of space for inline versions. I do this in the messages, assuming that changes to this schema will be handled by the outermost encoder.

What I may be missing in this case, is having the version passed all the way through from the topmost dencoder. I'll think about it more, thanks for the comment

Reply:
@batrick explained why we will have to add versioning to every separately encoded entity. Will have to do that